libraries/AP_NavEKF/AP_NavEKF_Source.cpp

```diff
@@ -0,0 +1,492 @@
+            return false;
+        }
+
+        // check velxy
+        switch ((SourceXY)_source_set[i].velxy.get()) {
+        case SourceXY::NONE:
+            break;
+        case SourceXY::GPS:
+            gps_required = true;
+            break;
+        case SourceXY::OPTFLOW:
+            optflow_required = true;
+            break;
+        case SourceXY::EXTNAV:
+            visualodom_required = true;
+            break;
+        case SourceXY::WHEEL_ENCODER:
+            // ToDo: add wheelencoder_required and test below
+            break;
+        case SourceXY::BEACON:
+        default:
+            // invalid velxy value
+            hal.util->snprintf(failure_msg, failure_msg_len, "Check EK3_SRC%d_VELXY", (int)i+1);
+            return false;
+        }
+
+        // check posz
+        switch ((SourceZ)_source_set[i].posz.get()) {
+        case SourceZ::BARO:
+            baro_required = true;
+            break;
+        case SourceZ::RANGEFINDER:
+            rangefinder_required = true;
+            break;
+        case SourceZ::GPS:
+            gps_required = true;
+            break;
+        case SourceZ::BEACON:
+            beacon_required = true;
+            break;
+        case SourceZ::EXTNAV:
+            visualodom_required = true;
+            break;
+        case SourceZ::NONE:
+        default:
+            // invalid posz value
+            hal.util->snprintf(failure_msg, failure_msg_len, "Check EK3_SRC%d_POSZ", (int)i+1);
+            return false;
+        }
+
+        // check velz
+        switch ((SourceZ)_source_set[i].velz.get()) {
+        case SourceZ::NONE:
+            break;
+        case SourceZ::GPS:
+            gps_required = true;
+            break;
+        case SourceZ::EXTNAV:
+            visualodom_required = true;
+            break;
+        case SourceZ::BARO:
+        case SourceZ::RANGEFINDER:
+        case SourceZ::BEACON:
+        default:
+            // invalid velz value
+            hal.util->snprintf(failure_msg, failure_msg_len, "Check EK3_SRC%d_VELZ", (int)i+1);
+            return false;
+        }
+
+        // check yaw
+        switch ((SourceYaw)_source_set[i].yaw.get()) {
+        case SourceYaw::NONE:
+        case SourceYaw::COMPASS:
+        case SourceYaw::EXTERNAL:
+        case SourceYaw::EXTERNAL_COMPASS_FALLBACK:
+            // valid yaw value
+            break;
+        default:
+            // invalid yaw value
+            hal.util->snprintf(failure_msg, failure_msg_len, "Check EK3_SRC%d_YAW", (int)i+1);
+            return false;
+        }
+    }
+
+    // check all required sensors are available
+    const char* ekf_requires_msg = "EK3 sources require %s";
+    if (baro_required && (dal.baro().num_instances() == 0)) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "Baro");
+        return false;
+    }
+
+    if (beacon_required && (dal.beacon() == nullptr || !dal.beacon()->enabled())) {
```

Comment:
yes, this is intentional.  The thought was that we don't know if the sensor can always provide data or not.  We can imagine that the landing pad has a beacon system but this system can only provide distances when the vehicle is within a certain distance.  While the vehicle is arming it may not be near the landing pad.

If we are actively using the sensor for position estimation then the existing EKF position checks should catch an issue with the sensor not working.

Txs for the detailed review.